test/idleness_test.go

```diff
@@ -0,0 +1,481 @@
+		defer sCancel()
+		if cc.WaitForStateChange(sCtx, connectivity.Ready) {
+			errCh <- fmt.Errorf("Connectivity state changed to %q when expected to stay in READY", cc.GetState())
+			return
+		}
+
+		// Verify that there are no idleness related channelz events.
+		if err := channelzTraceEventNotFound(ctx, "entering idle mode"); err != nil {
+			errCh <- err
+			return
+		}
+		if err := channelzTraceEventNotFound(ctx, "exiting idle mode"); err != nil {
+			errCh <- err
+			return
+		}
+
+		// Unblock the unary RPC on the server.
+		close(blockCh)
+		errCh <- nil
+	}()
+
+	// Make a unary RPC that blocks on the server, thereby ensuring that the
+	// count of active RPCs on the client is non-zero.
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Errorf("EmptyCall RPC failed: %v", err)
+	}
+
+	select {
+	case err := <-errCh:
+		if err != nil {
+			t.Fatal(err)
+		}
+	case <-ctx.Done():
+		t.Fatalf("Timeout when trying to verify that an active RPC keeps channel from moving to IDLE")
+	}
+}
+
+// Tests the case where channel idleness is enabled by passing a small value for
+// idle_timeout. Verifies that activity on a READY channel (frequent and short
+// RPCs) keeps it from moving to IDLE.
+func (s) TestChannelIdleness_Enabled_ActiveSinceLastCheck(t *testing.T) {
+	// Setup channelz for testing.
+	czCleanup := channelz.NewChannelzStorageForTesting()
+	t.Cleanup(func() { czCleanupWrapper(czCleanup, t) })
+
+	// Create a ClientConn with a short idle_timeout.
+	r := manual.NewBuilderWithScheme("whatever")
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithIdleTimeout(defaultTestShortIdleTimeout),
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"round_robin":{}}]}`),
+	}
+	cc, err := grpc.Dial(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.Dial() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// Start a test backend and push an address update via the resolver.
+	backend := stubserver.StartTestService(t, nil)
+	t.Cleanup(func() { backend.Stop() })
+	r.UpdateState(resolver.State{Addresses: []resolver.Address{{Addr: backend.Address}}})
+
+	// Veirfy that the ClientConn moves to READY.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	for state := cc.GetState(); state != connectivity.Ready; state = cc.GetState() {
+		if !cc.WaitForStateChange(ctx, state) {
+			t.Fatal("Timeout when waiting for channel to switch to READY")
+		}
+	}
+
+	// For a duration of three times the configured idle timeout, making RPCs
+	// every now and then and ensure that the channel does not move out of
+	// READY.
+	sCtx, sCancel := context.WithTimeout(ctx, 3*defaultTestShortIdleTimeout)
+	defer sCancel()
+	go func() {
+		for ; sCtx.Err() == nil; <-time.After(defaultTestShortTimeout) {
```

Comment:
Maybe `defaultTestShortIdleTimeout/2` for this instead?  Otherwise these things that are otherwise not dependent on each other are actually dependent on each other for this test's correctness.